osu.Game/Screens/SelectV2/Carousel.cs

```diff
@@ -0,0 +1,444 @@
+        /// Of note, if no matching item exists all items will be deselected while waiting for potential new item which matches.
+        /// </remarks>
+        public virtual object? CurrentSelection
+        {
+            get => currentSelection;
+            set
+            {
+                if (currentSelectionCarouselItem != null)
+                    currentSelectionCarouselItem.Selected.Value = false;
+
+                currentSelection = value;
+
+                currentSelectionCarouselItem = null;
+                currentSelectionYPosition = null;
+                updateSelection();
+            }
+        }
+
+        private List<CarouselItem>? displayedCarouselItems;
+
+        private readonly DoublePrecisionScroll scroll;
+
+        protected Carousel()
+        {
+            InternalChildren = new Drawable[]
+            {
+                new Box
+                {
+                    Colour = Color4.Black,
+                    RelativeSizeAxes = Axes.Both,
+                },
+                scroll = new DoublePrecisionScroll
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Masking = false,
+                }
+            };
+
+            Items.BindCollectionChanged((_, _) => QueueFilter());
+        }
+
+        /// <summary>
+        /// Queue an asynchronous filter operation.
+        /// </summary>
+        public void QueueFilter() => Scheduler.AddOnce(() => filterTask = performFilter());
+
+        /// <summary>
+        /// Create a drawable for the given carousel item so it can be displayed.
+        /// </summary>
+        /// <remarks>
+        /// For efficiency, it is recommended the drawables are retrieved from a <see cref="DrawablePool{T}"/>.
+        /// </remarks>
+        /// <param name="item">The item which should be represented by the returned drawable.</param>
+        /// <returns>The manifested drawable.</returns>
+        protected abstract Drawable GetDrawableForDisplay(CarouselItem item);
+
+        /// <summary>
+        /// Create an internal carousel representation for the provided model object.
+        /// </summary>
+        /// <param name="model">The model.</param>
+        /// <returns>A <see cref="CarouselItem"/> representing the model.</returns>
+        protected abstract CarouselItem CreateCarouselItemForModel(T model);
+
+        #region Filtering and display preparation
+
+        private Task filterTask = Task.CompletedTask;
+        private CancellationTokenSource cancellationSource = new CancellationTokenSource();
+
+        private async Task performFilter()
+        {
+            Debug.Assert(SynchronizationContext.Current != null);
+
+            var cts = new CancellationTokenSource();
+
+            lock (this)
+            {
+                cancellationSource.Cancel();
+                cancellationSource = cts;
+            }
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
```

Comment:
Not sure how much of this you want to keep, but this and the bunch of log statements below look potentially removable.

Reply:
Kinda want to keep so we can track performance when deployed and users report potential risk issues.

At least initially.